lib/rspec/rails/matchers/have_enqueued_mail.rb

```diff
@@ -0,0 +1,166 @@
+
+        def with(*args)
+          @mail_args = args
+          super(*mailer_args)
+        end
+
+        def matches?(block)
+          raise ArgumentError, 'have_enqueued_mail and enqueue_mail only work with block arguments' unless block.respond_to?(:call)
+          check_active_job_adapter
+          super
+        end
+
+        def failure_message
+          "expected to enqueue #{base_message}".tap do |msg|
+            msg << "\n#{unmatching_mail_jobs_message}" if unmatching_mail_jobs.any?
+          end
+        end
+
+        def failure_message_when_negated
+          "expected not to enqueue #{base_message}"
+        end
+
+        private
+
+        def base_message
+          "#{@mailer_class.name}.#{@method_name}".tap do |msg|
+            msg << " #{expected_count_message}"
+            msg << " with #{@mail_args}," if @mail_args.any?
+            msg << " on queue #{@queue}," if @queue
+            msg << " at #{@at.inspect}," if @at
+            msg << " but enqueued #{@matching_jobs.size}"
+          end
+        end
+
+        def expected_count_message
+          "#{message_expectation_modifier} #{@expected_number} #{@expected_number == 1 ? 'time' : 'times'}"
+        end
+
+        def mailer_args
+          base_args = [@mailer_class.name, @method_name.to_s, 'deliver_now']
+
+          if @mail_args.any?
+            base_args + @mail_args
+          else
+            mailer_method_arity = @mailer_class.instance_method(@method_name).arity
+
+            number_of_args = if mailer_method_arity < 0
+                               (mailer_method_arity + 1).abs
+                             else
+                               mailer_method_arity
+                             end
+
+            base_args + Array.new(number_of_args) { anything }
+          end
+        end
+
+        def check_active_job_adapter
+          return if ::ActiveJob::QueueAdapters::TestAdapter === ::ActiveJob::Base.queue_adapter
+          raise StandardError, "To use HaveEnqueuedMail matcher set `ActiveJob::Base.queue_adapter = :test`"
+        end
+
+        def unmatching_mail_jobs
+          @unmatching_jobs.select do |job|
+            job[:job] == mailer_job
+          end
+        end
+
+        def unmatching_mail_jobs_message
+          msg = "Queued deliveries:"
+
+          unmatching_mail_jobs.each do |job|
+            msg << "\n  #{mail_job_message(job)}"
+          end
+
+          msg
+        end
+
+        def mail_job_message(job)
+          mailer_method = job[:args][0..1].join('.')
+
+          mailer_args = job[:args][3..-1]
+          msg_parts = []
+          msg_parts << "with #{mailer_args}" if mailer_args.any?
+          msg_parts << "on queue #{job[:queue]}" if job[:queue] && job[:queue] != 'mailers'
+          msg_parts << "at #{Time.at(job[:at])}" if job[:at]
+
+          "#{mailer_method} #{msg_parts.join(', ')}".strip
+        end
+
+        def mailer_job
+          ActionMailer::DeliveryJob
+        end
```

Comment:
Not a blocker, but why not just use the constant name directly in places?

Reply:
I had the hardest time keeping the constant straight in my head.  I kept wanting to call it `ActionDelivery::MailerJob` or `ActionMailer::DeliverJob`.  So I guess out of frustration, I just made it a method 😅